src/smp/ipi.c

```diff
@@ -21,62 +21,72 @@
  * or this call will idle forever */
 void ipiStallCoreCallback(bool_t irqPath)
 {
-    if (clh_is_self_in_queue() && !irqPath) {
-        /* The current thread is running as we would replace this thread with an idle thread
-         *
-         * The instruction should be re-executed if we are in kernel to handle syscalls.
-         * Also, thread in 'ThreadState_RunningVM' should remain in same state.
-         * Note that, 'ThreadState_Restart' does not always result in regenerating exception
-         * if we are in kernel to handle them, e.g. hardware single step exception. */
-        if (thread_state_ptr_get_tsType(&NODE_STATE(ksCurThread)->tcbState) == ThreadState_Running) {
-            setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
-        }
-
+    // ipiStallCore can only be called for a single core at a time to be able to assume
+    // kernel mutual exclusion as the caller node is blocked until this stall completes
+    // and all other nodes are out of the kernel or blocked on the lock.
+    assert(totalCoreBarrier == 1);
+
+    // There's 3 ways that this function can be called:
+    // - On the IRQ path where the lock isn't acquired
+    // - On the IRQ path where the lock is attempted to be acquired
+    // - On the non-IRQ path where the lock is attempted to be acquired
+    // !irqPath implies clh_is_self_in_queue()
+    assert(clh_is_self_in_queue() || irqPath);
+
+    /* The current thread is running as we would replace this thread with an idle thread
+     *
+     * The instruction should be re-executed if we are in kernel to handle syscalls.
+     * Also, thread in 'ThreadState_RunningVM' should remain in same state.
+     * Note that, 'ThreadState_Restart' does not always result in regenerating exception
+     * if we are in kernel to handle them, e.g. hardware single step exception. */
+    if (!irqPath && thread_state_ptr_get_tsType(&NODE_STATE(ksCurThread)->tcbState) == ThreadState_Running) {
+        setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
+    }
+#ifdef CONFIG_KERNEL_MCS
+    updateTimestamp();
+#endif
+    if (config_ternary(CONFIG_KERNEL_MCS, checkBudget(),1)) {
         SCHED_ENQUEUE_CURRENT_TCB;
-        switchToIdleThread();
+    }
+    switchToIdleThread();
+    NODE_STATE(ksSchedulerAction) = SchedulerAction_ResumeCurrentThread;
+    doMaskReschedule(ARCH_NODE_STATE(ipiReschedulePending));
```

Comment:
Triggering an IPI within an IPI handler is getting pretty sketchy... I believe this is currently needed because a timeoutFault generated by the current thread being at the end of its budget can make a thread runnable on a different node. As doMaskReschedule is a non-blocking IPI it only delivers an interrupt to the other nodes which only act on it when they independently get the lock.  (Compared with blocking IPIs which are executed by the target nodes without acquiring the lock). I expect that this means that if the reschedule IPI is sent back to the original node that triggered this synchronous IPI it will be delivered once that node releases the lock.

Reply:
Why not let the future schedule call do the `doMaskReschedule`? Sending reschedule IPIs is non-blocking, but receiving isn't. As the reschedule IPI is handled with the kernel lock taken, doing it here will force a third remote core to block on the kernel lock immediately.